mrmustard/lab_dev/states/dm.py

```diff
@@ -329,6 +330,300 @@ def normalize(self) -> DM:
         """
         return self / self.probability
 
+    def formal_stellar_decomposition(self, core_modes: Collection[int]) -> tuple[DM, Map]:
+        r"""
+        Computes the formal stellar decomposition for the DM.
+
+        Args:
+            core_modes: The set of modes defining core variables.
+
+        Returns:
+            core: The core state (`DM`)
+            phi: The Gaussian `Map` performing the stellar decomposition (not necessarily CPTP).
+
+        Note:
+            This method pulls out the map ``phi`` from the given state on the given modes, so that
+            the remaining state is a core state. Formally, we have
+            .. math::
+
+                \rho = (\phi\otimes\mathcal I) \rho_{\mathrm{core}}
+
+            where the map :math:`phi` acts on the given `core_modes` only.
+            Core states have favorable properties in the Fock representation
+            e.g., being sparse.
+
+        .. code-block::
+            >>> from mrmustard.lab_dev import DM, Vacuum
+
+            >>> rho = DM.random([0,1])
+            >>> core, phi = rho.formal_stellar_decomposition([0])
+            >>> assert (core >> Vacuum(1).dual).normalize() == Vacuum(0).dm()
+            >>> assert rho == core >> phi
+            >>> assert (core >> Vacuum(1).dual).normalize() == Vacuum(0).dm()
+        """
+        other_modes = [m for m in self.modes if m not in core_modes]
+        core_indices = self.wires[core_modes].indices
+        other_indices = self.wires[other_modes].indices
+        new_order = core_indices + other_indices
+        A, b, c = self.ansatz.reorder(new_order).triple
+
+        M = len(core_modes)
+
+        Am = A[: 2 * M, : 2 * M]
```

Comment:
can you update all of these indexing calls to include `...` to handle batch dims? i.e. 
```suggestion
        Am = A[..., : 2 * M, : 2 * M]
```

Reply:
Sure. Is it safe to call `math.block()` on these?

Reply:
Yeah it should be able to handle it